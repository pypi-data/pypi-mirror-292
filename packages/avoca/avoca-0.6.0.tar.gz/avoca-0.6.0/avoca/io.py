from pathlib import Path

import pandas as pd


def to_csv(df: pd.DataFrame, path: Path, **kwargs) -> None:
    """Export a dataframe to a csv file."""
    df.to_csv(path, index=False, **kwargs)


def from_csv(path: Path) -> pd.DataFrame:
    """Read a csv file to a dataframe."""
    df = pd.read_csv(path, header=[0, 1])
    parse_dates = [("-", "datetime"), ("-", "start_datetime"), ("-", "end_datetime")]
    # Convert the datetime column to a datetime object
    for col in parse_dates:
        if col not in df.columns:
            continue
        df[col] = pd.to_datetime(df[col])
    return df
