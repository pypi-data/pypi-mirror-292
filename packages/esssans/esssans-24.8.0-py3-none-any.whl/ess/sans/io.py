# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)

import scipp as sc
import scippnexus as snx
from scippnexus.application_definitions import nxcansas

from .types import BackgroundSubtractedIofQ, OutFilename, RunNumber, RunTitle


def save_background_subtracted_iofq(
    *,
    iofq: BackgroundSubtractedIofQ,
    out_filename: OutFilename,
    run_number: RunNumber,
    run_title: RunTitle,
) -> None:
    """Save background-subtracted I(Q) histogram as an NXcanSAS file."""
    if iofq.bins is None:
        da = iofq.copy(deep=False)
    else:
        da = iofq.hist()
    if da.coords.is_edges('Q'):
        da.coords['Q'] = sc.midpoints(da.coords['Q'])
    with snx.File(out_filename, 'w') as f:
        f['sasentry'] = nxcansas.SASentry(title=run_title, run=run_number)
        f['sasentry']['sasdata'] = nxcansas.SASdata(da, Q_variances='resolutions')
