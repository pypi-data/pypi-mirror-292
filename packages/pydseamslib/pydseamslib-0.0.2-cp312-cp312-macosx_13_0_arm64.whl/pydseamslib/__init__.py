from pydseamslib import cyoda

__all__ = ["cyoda"]
