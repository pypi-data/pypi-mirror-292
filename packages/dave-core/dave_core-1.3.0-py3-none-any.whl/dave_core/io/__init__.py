# Copyright (c) 2022-2024 by Fraunhofer Institute for Energy Economics and Energy System Technology (IEE)
# Kassel and individual contributors (see AUTHORS file for details). All rights reserved.
# Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.

from dave_core.io.convert_format import *
from dave_core.io.file_io import *
from dave_core.io.io_utils import *
