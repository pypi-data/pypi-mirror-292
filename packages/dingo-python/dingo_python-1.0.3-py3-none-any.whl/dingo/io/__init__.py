from dingo.io.input import InputModel, RawInputModel
from dingo.io.output import SummaryModel, ErrorInfo
