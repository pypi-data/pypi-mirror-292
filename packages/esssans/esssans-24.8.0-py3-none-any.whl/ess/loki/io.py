# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
"""
Loading and merging of LoKI data.
"""

import scipp as sc
from ess.reduce import nexus

from ess.sans.types import (
    DetectorEventData,
    Filename,
    MonitorEventData,
    MonitorType,
    NeXusDetector,
    NeXusDetectorName,
    NeXusMonitor,
    NeXusMonitorName,
    RawSample,
    RawSource,
    RunType,
)


def load_nexus_sample(file_path: Filename[RunType]) -> RawSample[RunType]:
    return RawSample[RunType](nexus.load_sample(file_path))


def dummy_load_sample(file_path: Filename[RunType]) -> RawSample[RunType]:
    return RawSample[RunType](
        sc.DataGroup({'position': sc.vector(value=[0, 0, 0], unit='m')})
    )


def load_nexus_source(file_path: Filename[RunType]) -> RawSource[RunType]:
    return RawSource[RunType](nexus.load_source(file_path))


def load_nexus_detector(
    file_path: Filename[RunType], detector_name: NeXusDetectorName
) -> NeXusDetector[RunType]:
    # Events will be loaded later. Should we set something else as data instead, or
    # use different NeXus definitions to completely bypass the (empty) event load?
    dg = nexus.load_detector(
        file_path=file_path,
        detector_name=detector_name,
        selection={'event_time_zero': slice(0, 0)},
    )
    # The name is required later, e.g., for determining logical detector shape
    dg['detector_name'] = detector_name
    return NeXusDetector[RunType](dg)


def load_nexus_monitor(
    file_path: Filename[RunType], monitor_name: NeXusMonitorName[MonitorType]
) -> NeXusMonitor[RunType, MonitorType]:
    return NeXusMonitor[RunType, MonitorType](
        nexus.load_monitor(
            file_path=file_path,
            monitor_name=monitor_name,
            selection={'event_time_zero': slice(0, 0)},
        )
    )


def load_detector_event_data(
    file_path: Filename[RunType], detector_name: NeXusDetectorName
) -> DetectorEventData[RunType]:
    da = nexus.load_event_data(file_path=file_path, component_name=detector_name)
    return DetectorEventData[RunType](da)


def load_monitor_event_data(
    file_path: Filename[RunType], monitor_name: NeXusMonitorName[MonitorType]
) -> MonitorEventData[RunType, MonitorType]:
    da = nexus.load_event_data(file_path=file_path, component_name=monitor_name)
    return MonitorEventData[RunType, MonitorType](da)


providers = (
    load_nexus_detector,
    load_nexus_monitor,
    load_nexus_sample,
    load_nexus_source,
    load_detector_event_data,
    load_monitor_event_data,
)
"""Providers for loading single files."""
