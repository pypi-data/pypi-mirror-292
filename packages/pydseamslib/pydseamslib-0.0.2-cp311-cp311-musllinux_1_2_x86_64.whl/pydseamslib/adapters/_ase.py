from pydseamslib.cyoda import PointDouble, PointCloudDouble
import numpy as np
import ase


def map_LAMMPS_IDs_to_atomic_symbols(dict_map: dict(), atms_a: ase.Atoms):
    # Traversing by atom is faster than traversing atoms by key value in the map,
    # since there will be fewer map values than atoms
    # See https://github.com/d-SEAMS/PydSEAMSlib/pull/11#issuecomment-2243986351
    # for benchmarks
    for pt in atms_a:
        if pt.number in dict_map.keys():
            pt.symbol = dict_map[pt.number]
    return atms_a


def swap_key_value(my_dict_a):
    new_dict = {value: key for key, value in my_dict_a.items()}
    return new_dict


def list_pts(atms: ase.Atoms, lmp_a: dict, splicemask: list[bool], molids: list[int]):
    ptslist = []
    inslice_atms = atms[splicemask]
    map_lmp = swap_key_value(lmp_a)
    all_atm_id = np.asarray([x.index + 1 for x in atms])[splicemask]
    for idx, atm in enumerate(inslice_atms):
        pd = PointDouble()
        pd.x = atm.position[0]
        pd.y = atm.position[1]
        pd.z = atm.position[2]
        pd.c_type = map_lmp[atm.symbol]
        pd.inSlice = True  # Assumes that the atom provided are in the Slice.
        pd.atomID = all_atm_id[idx]
        pd.molID = molids[idx]
        ptslist.append(pd)
    return ptslist


def to_pointcloud(
    atms_a: ase.Atoms,
    lmp_a: dict,
    splicemask: list[bool],
    molids: list[int],
    fname,
    currentFrame=[1],
):
    _pcd = PointCloudDouble()
    inslice_atms = atms_a[splicemask]
    # TODO(ruhila): Assumes a rectangular box
    _pcd.box = np.diag(np.asarray(inslice_atms.get_cell()))
    _pcd.nop = len(inslice_atms)
    _pcd.pts = list_pts(atms_a, lmp_a, splicemask, molids)
    _pcd.idIndexMap = make_indexmap(atms_a, splicemask)
    _pcd.currentFrame = 1
    _pcd.boxLow = boxlow(fname)
    return _pcd


def make_indexmap(atms: ase.Atoms, splicemask: list[bool]):
    inslice_atms = atms[splicemask]
    atm_id = np.asarray([x.index + 1 for x in atms])[splicemask]
    ret = dict()
    for idx in range(len(inslice_atms)):
        key = int(atm_id[idx])
        ret[key] = idx
    return ret


def boxlow(fname):
    # Read the contents of the file into a variable
    opfile = open(fname)
    data = opfile.read()
    filelist = []
    for i in range(5, 8):
        parfile = data.splitlines()[i]
        needfile = parfile.split()[0]
        floatfile = float(needfile)
        filelist.append(floatfile)
    # Don't forget to close the file again
    opfile.close()
    return filelist
