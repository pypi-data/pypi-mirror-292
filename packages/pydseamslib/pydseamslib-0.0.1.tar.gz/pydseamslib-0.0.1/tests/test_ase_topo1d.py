import pprint
import shutil
import numpy as np

from approvaltests import verify, verify_file
from approvaltests.namer.default_namer_factory import NamerFactory
from ase.io import read as aseread


from pydseamslib import cyoda
from pydseamslib.adapters import _ase
from pathlib import Path

TRAJ = Path("subprojects/seams-core/input/traj/exampleTraj.lammpstrj")
strTRJ = str(TRAJ.absolute())


def test_nlist():
    # Get the frame
    resCloud = cyoda.readLammpsTrjreduced(
        filename=strTRJ,
        targetFrame=1,
        typeI=2,  # oxygenAtomType
        isSlice=False,
        coordLow=[0, 0, 0],
        coordHigh=[0, 0, 0],
    )

    # Construct a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )
    assert pcd.box == resCloud.box
    assert pcd.nop == resCloud.nop
    assert pcd.idIndexMap == resCloud.idIndexMap
    assert pcd.currentFrame == resCloud.currentFrame
    assert pcd.boxLow == resCloud.boxLow

    for idx in range(len(pcd.pts)):
        assert pcd.pts[idx].x == resCloud.pts[idx].x
        assert pcd.pts[idx].y == resCloud.pts[idx].y
        assert pcd.pts[idx].z == resCloud.pts[idx].z
        assert pcd.pts[idx].c_type == resCloud.pts[idx].c_type
        assert pcd.pts[idx].inSlice == resCloud.pts[idx].inSlice
        assert pcd.pts[idx].atomID == resCloud.pts[idx].atomID
        assert pcd.pts[idx].molID == resCloud.pts[idx].molID

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Validate the results
    verify(pprint.pformat(nl))


def test_hbnlist():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )

    # Validate the results
    verify(pprint.pformat(hl))


def test_hbnlist1():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )

    # Validate the results
    verify(pprint.pformat(hL))


def test_rings():
    # Get the frame, make a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )

    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )

    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )
    # Gets every ring (non-primitives included)
    Rgs = cyoda.ringNetwork(
        nList=hL,
        maxDepth=6,
    )

    # Validate the results
    verify(pprint.pformat(Rgs))


def test_prisms():
    # Construct a pointcloud
    atms = aseread(strTRJ)
    # TODO(ruhi): How should these be passed
    lammps_to_ase = {1: "H", 2: "O"}
    atms = _ase.map_LAMMPS_IDs_to_atomic_symbols(lammps_to_ase, atms)
    only_O_mask = [x.symbol == "O" for x in atms]
    molOID = np.repeat(np.arange(1, sum(only_O_mask) + 1), 1)
    pcd = _ase.to_pointcloud(
        atms, lammps_to_ase, only_O_mask, molOID, TRAJ, currentFrame=[1]
    )
    # Calculate the neighborlist by ID
    nl = cyoda.neighListO(
        rcutoff=3.5,
        yCloud=pcd,
        typeI=2,  # oxygenAtomType
    )
    # Get the hydrogen-bonded network for the current frame
    hl = cyoda.populateHbonds(
        filename=strTRJ,
        yCloud=pcd,
        nList=nl,
        targetFrame=1,
        Htype=1,  # hydrogen atom type
    )
    # Hydrogen-bonded network using indices not IDs
    hL = cyoda.neighbourListByIndex(
        yCloud=pcd,
        nList=hl,
    )
    # Gets every ring (non-primitives included)
    Rgs = cyoda.ringNetwork(
        nList=hL,
        maxDepth=6,
    )
    # Ensure the directory is not present before beginning
    outDir = "runOne/"  # / is important for the C++ engine..
    if Path(outDir).exists():
        shutil.rmtree(outDir)
    # Does the prism analysis for quasi-one-dimensional ice
    cyoda.prismAnalysis(
        path=outDir,
        rings=Rgs,
        nList=hL,
        yCloud=pcd,
        maxDepth=6,
        atomID=0,
        firstFrame=1,  # targetFrame
        currentFrame=1,  # frame
        doShapeMatching=False,
    )
    # Validate the run results
    verify_file(
        Path(f"{outDir}/topoINT/dataFiles/system-prisms-1.data"),
        options=NamerFactory.with_parameters("systemPrisms"),
    )
    verify_file(
        Path(f"{outDir}/topoINT/nPrisms.dat"),
        options=NamerFactory.with_parameters("nPrisms"),
    )
