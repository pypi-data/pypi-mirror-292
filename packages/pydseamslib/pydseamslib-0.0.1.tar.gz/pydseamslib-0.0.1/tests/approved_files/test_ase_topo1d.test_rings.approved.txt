[[0, 51, 54, 183, 170, 144],
 [0, 51, 54, 81],
 [0, 51, 60, 144],
 [0, 51, 193, 93, 5, 144],
 [0, 51, 193, 242, 5, 144],
 [0, 81, 31, 23, 170, 144],
 [0, 81, 54, 183, 60, 144],
 [0, 81, 54, 183, 170, 144],
 [1, 89, 150, 215, 138, 219],
 [1, 89, 150, 223],
 [1, 89, 148, 219],
 [1, 89, 148, 215, 138, 223],
 [1, 223, 150, 215, 148, 219],
 [1, 223, 29, 10, 117, 219],
 [1, 223, 29, 238, 117, 219],
 [1, 223, 138, 219],
 [2, 24, 111, 17],
 [3, 107, 73, 235],
 [3, 107, 73, 121, 152, 97],
 [3, 107, 158, 97],
 [3, 107, 160, 115],
 [3, 107, 160, 34, 203, 97],
 [3, 115, 203, 97],
 [3, 97, 152, 235],
 [4, 220, 142, 241],
 [4, 220, 198, 46, 229, 241],
 [4, 220, 198, 187, 229, 241],
 [5, 144, 60, 51, 193, 93],
 [5, 144, 60, 51, 193, 242],
 [5, 242, 175, 62],
 [5, 242, 193, 93],
 [5, 62, 56, 93],
 [5, 62, 175, 192, 193, 93],
 [6, 20, 249, 74],
 [6, 20, 249, 178, 15, 30],
 [6, 20, 94, 30],
 [6, 20, 94, 178, 15, 74],
 [6, 30, 15, 71, 92, 11],
 [6, 30, 15, 74],
 [6, 30, 131, 11],
 [6, 30, 131, 71, 92, 74],
 [6, 30, 94, 178, 249, 74],
 [6, 11, 131, 71, 15, 74],
 [6, 11, 92, 74],
 [7, 88, 210, 167],
 [7, 88, 210, 168, 65, 101],
 [7, 88, 212, 101],
 [7, 88, 212, 168, 65, 167],
 [7, 101, 65, 167],
 [7, 101, 212, 168, 210, 167],
 [8, 63, 155, 45, 237, 206],
 [8, 63, 155, 68],
 [8, 63, 130, 45, 237, 68],
 [8, 63, 130, 206],
 [8, 206, 130, 45, 155, 68],
 [8, 206, 145, 66, 227, 68],
 [8, 206, 145, 104, 227, 68],
 [8, 206, 237, 68],
 [9, 234, 77, 135],
 [9, 234, 205, 48, 176, 135],
 [10, 29, 59, 146],
 [10, 29, 59, 164, 27, 117],
 [10, 29, 238, 164, 27, 146],
 [10, 29, 238, 117],
 [10, 29, 223, 138],
 [10, 138, 219, 117],
 [10, 117, 27, 146],
 [10, 117, 238, 164, 59, 146],
 [11, 131, 30, 15, 74, 92],
 [11, 131, 41, 19],
 [11, 131, 41, 140, 156, 92],
 [11, 131, 71, 92],
 [11, 131, 71, 140, 156, 19],
 [11, 19, 156, 92],
 [11, 19, 41, 140, 71, 92],
 [12, 141, 146, 59, 164, 55],
 [12, 141, 146, 27],
 [12, 141, 207, 59, 164, 27],
 [12, 141, 207, 55],
 [12, 141, 207, 67, 214, 161],
 [12, 141, 186, 67, 214, 55],
 [12, 141, 186, 161],
 [12, 161, 214, 55],
 [12, 161, 186, 67, 207, 55],
 [12, 27, 164, 55],
 [12, 27, 146, 59, 207, 55],
 [13, 110, 32, 247],
 [13, 110, 32, 244, 113, 188],
 [13, 110, 33, 188],
 [13, 110, 33, 244, 113, 247],
 [13, 247, 32, 244, 33, 188],
 [13, 247, 103, 86],
 [13, 247, 103, 127, 126, 188],
 [13, 247, 113, 188],
 [13, 247, 113, 127, 126, 86],
 [13, 86, 103, 127, 113, 188],
 [13, 86, 126, 188],
 [14, 38, 185, 120, 96, 149],
 [14, 38, 185, 236, 208],
 [14, 38, 230, 123, 208],
 [14, 38, 154, 149],
 [15, 71, 92, 74],
 [15, 71, 131, 30],
 [15, 74, 249, 20, 94, 30],
 [15, 74, 249, 178],
 [15, 30, 94, 178],
 [16, 21, 128, 163],
 [16, 21, 128, 216, 114, 99],
 [16, 21, 167, 210],
 [16, 21, 239, 99],
 [16, 21, 239, 216, 114, 163],
 [16, 163, 128, 216, 239, 99],
 [16, 163, 114, 99],
 [17, 176, 48],
 [17, 176, 135, 111],
 [17, 48, 226, 202, 111],
 [18, 184, 22, 190],
 [18, 184, 22, 78, 58, 40],
 [18, 184, 181, 40],
 [18, 184, 181, 78, 58, 190],
 [18, 190, 22, 78, 181, 40],
 [18, 190, 58, 40],
 [18, 190, 58, 233, 133, 129],
 [18, 190, 172, 129],
 [18, 190, 172, 233, 133, 40],
 [18, 40, 133, 129],
 [18, 40, 58, 233, 172, 129],
 [19, 156, 140, 41],
 [19, 156, 92, 71, 131, 41],
 [19, 156, 125, 85],
 [19, 156, 125, 108, 43, 41],
 [19, 41, 43, 85],
 [20, 245, 240, 213, 178, 94],
 [20, 245, 240, 249],
 [20, 245, 112, 94],
 [20, 245, 112, 213, 178, 249],
 [20, 249, 178, 94],
 [20, 249, 240, 213, 112, 94],
 [21, 128, 163, 114, 99, 239],
 [21, 128, 216, 239],
 [22, 184, 130, 45, 155, 78],
 [22, 184, 130, 63],
 [22, 184, 181, 45, 155, 63],
 [22, 184, 181, 40, 58, 190],
 [22, 184, 181, 78],
 [22, 190, 58, 78],
 [22, 63, 155, 78],
 [22, 63, 130, 45, 181, 78],
 [23, 121, 73, 31],
 [23, 121, 73, 235, 28, 209],
 [23, 121, 152, 209],
 [23, 121, 152, 235, 28, 31],
 [23, 209, 152, 235, 73, 31],
 [23, 209, 28, 31],
 [23, 209, 183, 54, 81, 31],
 [23, 209, 183, 170],
 [23, 31, 81, 54, 183, 170],
 [24, 221, 208, 236],
 [24, 221, 202, 111],
 [25, 143, 116, 222, 69, 95],
 [25, 143, 116, 171],
 [25, 143, 147, 44],
 [25, 143, 147, 196, 26, 95],
 [25, 143, 189, 95],
 [25, 143, 189, 196, 26, 44],
 [25, 143, 189, 222, 69, 171],
 [25, 171, 69, 95],
 [25, 171, 116, 222, 189, 95],
 [25, 44, 26, 95],
 [25, 44, 147, 196, 189, 95],
 [26, 95, 189, 143, 147, 44],
 [26, 95, 189, 196],
 [26, 196, 72, 42, 70, 44],
 [26, 196, 72, 162],
 [26, 196, 147, 42, 70, 162],
 [26, 196, 147, 44],
 [26, 44, 70, 162],
 [26, 44, 147, 42, 72, 162],
 [27, 164, 55, 207, 141, 146],
 [27, 164, 59, 146],
 [27, 164, 238, 117],
 [27, 117, 238, 29, 59, 146],
 [28, 31, 81, 54, 183, 209],
 [28, 31, 73, 235],
 [28, 31, 73, 121, 152, 209],
 [28, 235, 152, 209],
 [29, 59, 164, 238],
 [29, 238, 117, 219, 138, 223],
 [32, 110, 33, 188, 113, 247],
 [32, 110, 33, 244],
 [32, 110, 33, 225, 243, 106],
 [32, 110, 79, 106],
 [32, 247, 113, 244],
 [32, 106, 79, 225, 33, 244],
 [32, 106, 243, 225, 33, 244],
 [33, 188, 113, 244],
 [33, 110, 79, 225],
 [34, 203, 97, 158, 107, 160],
 [34, 203, 115, 160],
 [35, 82, 100, 194, 50, 53],
 [35, 82, 100, 194, 50, 218],
 [35, 82, 100, 159, 232, 53],
 [35, 82, 100, 159, 232, 37],
 [35, 82, 118, 37],
 [35, 82, 118, 159, 232, 53],
 [35, 82, 177, 194, 50, 53],
 [35, 82, 177, 218],
 [35, 218, 50, 53],
 [35, 37, 232, 53],
 [36, 61, 86, 103],
 [36, 61, 86, 126, 127, 191],
 [36, 61, 246, 126, 127, 103],
 [36, 61, 246, 191],
 [36, 61, 76, 90],
 [36, 61, 76, 180, 102, 191],
 [36, 90, 102, 191],
 [36, 103, 86, 126, 246, 191],
 [36, 103, 127, 191],
 [37, 118, 75, 109, 139, 232],
 [37, 118, 75, 80],
 [37, 118, 159, 109, 139, 80],
 [37, 118, 159, 232],
 [37, 80, 75, 109, 159, 232],
 [37, 80, 139, 232],
 [38, 185, 120, 154],
 [39, 182, 91, 49, 52, 87],
 [39, 182, 91, 49, 98, 87],
 [39, 182, 91, 134],
 [39, 197, 157, 87],
 [39, 197, 228, 83, 98, 87],
 [39, 87, 52, 134],
 [39, 87, 98, 49, 91, 134],
 [40, 133, 129, 172, 190, 58],
 [40, 133, 233, 58],
 [40, 58, 78, 181],
 [41, 43, 85, 125, 156, 140],
 [41, 43, 108, 125, 156, 140],
 [41, 131, 71, 140],
 [42, 72, 162, 70],
 [42, 72, 166, 137, 47, 70],
 [42, 72, 166, 124, 47, 70],
 [42, 72, 196, 147],
 [42, 70, 44, 147],
 [43, 85, 125, 108],
 [43, 85, 125, 195, 122, 153],
 [43, 85, 169, 153],
 [43, 85, 169, 195, 122, 108],
 [43, 153, 122, 108],
 [43, 153, 169, 195, 125, 108],
 [44, 70, 162, 72, 196, 147],
 [45, 130, 63, 155],
 [45, 130, 206, 237],
 [45, 130, 184, 181],
 [45, 181, 78, 155],
 [45, 155, 68, 237],
 [46, 142, 241, 229],
 [46, 142, 220, 198],
 [46, 198, 187, 229],
 [46, 198, 199, 151, 224, 229],
 [47, 70, 162, 72, 166, 137],
 [47, 70, 162, 124],
 [47, 137, 77, 165],
 [47, 137, 166, 124],
 [47, 124, 166, 234, 77, 165],
 [48, 176, 135, 77, 234, 205],
 [48, 176, 165, 77, 234, 205],
 [49, 52, 87, 98],
 [49, 52, 134, 91],
 [49, 52, 134, 148, 215, 150],
 [49, 52, 89, 150],
 [49, 150, 89, 148, 134, 91],
 [49, 150, 215, 148, 134, 91],
 [50, 53, 232, 159, 100, 194],
 [50, 218, 177, 173, 200, 204],
 [50, 218, 177, 194],
 [50, 218, 179, 204],
 [50, 194, 177, 173, 200, 204],
 [50, 194, 177, 173, 179, 204],
 [51, 54, 183, 60],
 [52, 134, 148, 89],
 [55, 164, 59, 207],
 [55, 207, 141, 186, 161, 214],
 [55, 207, 67, 214],
 [56, 220, 142, 62],
 [56, 62, 175, 192, 193, 93],
 [56, 62, 175, 242, 193, 93],
 [57, 84, 114, 163],
 [57, 84, 211, 157, 83, 228],
 [57, 84, 211, 136],
 [57, 84, 211, 216, 128, 163],
 [57, 84, 197, 157, 83, 136],
 [57, 84, 197, 228],
 [57, 228, 197, 157, 211, 136],
 [57, 228, 83, 136],
 [57, 136, 211, 216, 114, 163],
 [57, 136, 211, 216, 128, 163],
 [58, 190, 172, 233],
 [59, 146, 141, 207],
 [60, 144, 170, 183],
 [61, 86, 103, 127, 191, 246],
 [61, 86, 126, 246],
 [61, 246, 191, 102, 90, 76],
 [61, 246, 191, 102, 180, 76],
 [63, 155, 78, 181, 184, 130],
 [63, 155, 68, 237, 206, 130],
 [64, 169, 153, 122, 105, 201],
 [64, 169, 153, 119],
 [64, 169, 195, 122, 105, 119],
 [64, 169, 195, 201],
 [64, 201, 195, 122, 153, 119],
 [64, 201, 105, 119],
 [64, 201, 173, 200, 204, 119],
 [64, 201, 173, 179, 204, 119],
 [65, 101, 212, 88, 210, 167],
 [65, 101, 212, 168],
 [65, 101, 212, 217, 248, 132],
 [65, 101, 231, 132],
 [65, 101, 231, 217, 248, 168],
 [65, 167, 210, 168],
 [65, 132, 231, 217, 212, 168],
 [65, 132, 248, 168],
 [66, 145, 96, 149],
 [66, 145, 96, 120, 154, 227],
 [66, 145, 104, 227],
 [66, 145, 104, 120, 154, 149],
 [66, 145, 206, 237],
 [66, 237, 68, 227],
 [66, 149, 96, 120, 104, 227],
 [66, 149, 154, 227],
 [67, 80, 75, 109, 161, 214],
 [67, 80, 75, 186],
 [67, 80, 139, 109, 161, 214],
 [67, 80, 139, 109, 161, 186],
 [67, 186, 141, 207],
 [67, 186, 161, 214],
 [68, 227, 104, 145, 206, 237],
 [69, 171, 116, 143, 189, 95],
 [69, 171, 116, 222],
 [69, 95, 189, 222],
 [71, 92, 156, 140],
 [72, 162, 124, 166],
 [73, 235, 152, 121],
 [73, 235, 152, 97, 158, 107],
 [73, 107, 158, 97, 152, 121],
 [75, 109, 139, 80],
 [75, 109, 159, 118],
 [75, 109, 161, 186],
 [75, 80, 139, 232, 159, 118],
 [76, 90, 102, 180],
 [76, 90, 102, 133, 129, 172],
 [76, 90, 233, 133, 129, 180],
 [76, 90, 233, 172],
 [76, 172, 129, 180],
 [76, 172, 233, 133, 102, 180],
 [77, 135, 176, 165],
 [77, 137, 166, 234],
 [79, 106, 213, 112],
 [79, 106, 213, 240, 245, 225],
 [79, 106, 243, 240, 245, 112],
 [79, 106, 243, 225],
 [79, 112, 213, 240, 243, 225],
 [79, 112, 245, 225],
 [82, 100, 194, 177],
 [82, 100, 159, 118],
 [83, 98, 87, 157],
 [83, 157, 197, 228],
 [83, 157, 211, 136],
 [83, 136, 211, 84, 197, 228],
 [84, 114, 216, 211],
 [84, 211, 157, 197],
 [85, 125, 108, 122, 153, 169],
 [85, 125, 195, 169],
 [86, 103, 127, 126],
 [86, 103, 247, 113, 188, 126],
 [88, 210, 168, 212],
 [89, 150, 215, 148],
 [89, 150, 223, 138, 219, 148],
 [90, 102, 180, 129, 172, 233],
 [90, 102, 133, 233],
 [94, 112, 213, 178],
 [94, 112, 245, 240, 249, 178],
 [96, 149, 154, 120],
 [96, 149, 154, 227, 104, 145],
 [96, 120, 104, 145],
 [97, 203, 115, 160, 107, 158],
 [99, 239, 216, 114],
 [101, 212, 168, 248, 132, 231],
 [101, 212, 217, 231],
 [102, 180, 129, 133],
 [103, 127, 113, 247],
 [104, 227, 154, 120],
 [105, 119, 204, 200],
 [105, 119, 204, 179, 173, 201],
 [105, 119, 153, 122],
 [105, 119, 153, 169, 195, 201],
 [105, 122, 195, 201],
 [105, 200, 173, 201],
 [106, 213, 112, 245, 225, 243],
 [106, 213, 240, 243],
 [108, 122, 195, 125],
 [109, 139, 232, 159],
 [112, 213, 240, 245],
 [113, 188, 126, 127],
 [114, 216, 128, 163],
 [115, 116, 171, 160],
 [116, 143, 189, 222],
 [122, 153, 169, 195],
 [123, 208, 221, 202, 226],
 [126, 246, 191, 127],
 [129, 133, 233, 172],
 [132, 231, 151, 199, 174, 248],
 [132, 231, 217, 248],
 [138, 215, 150, 223],
 [138, 215, 148, 219],
 [142, 241, 229, 187, 198, 220],
 [143, 147, 196, 189],
 [151, 224, 229, 187, 174, 199],
 [151, 224, 229, 187, 198, 199],
 [151, 199, 217, 231],
 [168, 248, 217, 212],
 [173, 177, 218, 179],
 [173, 200, 204, 179],
 [174, 187, 198, 199],
 [174, 199, 217, 248],
 [175, 192, 193, 242],
 [178, 213, 240, 249],
 [225, 243, 240, 245]]
